### PR TITLE
Implement IPC call_w_str, vtable for external package and interface query. 

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

**Review by danewalton — COMMENTED** [review]
unfortunately don't have the cycles right now for a full in depth review but left some thoughts

File: inc/_az_ulib_ipc_query.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+#ifndef _az_ULIB_IPC_QUERY_INTERFACE_H
+#define _az_ULIB_IPC_QUERY_INTERFACE_H
+
+#include "az_ulib_result.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#else
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/_az_ulib_ipc_query.h:13]: No, removed.

File: inc/_az_ulib_ipc_query.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#ifndef _az_ULIB_IPC_QUERY_INTERFACE_H
```
Comment (mamokarz) [inline inc/_az_ulib_ipc_query.h:5]: >_az_ULIB_IPC_QUERY_INTERFACE_H [](http://example.com/codeflow?start=8&length=30)

_az_ULIB_IPC_QUERY_H

File: inc/az_ulib_descriptor_api.h [hunk 3]
```diff
@@ -41,25 +41,39 @@
 typedef struct az_ulib_capability_descriptor_tag
 {
   /** The `\0` terminated `const az_span` with the capability name. */
-  const az_span name;
+  const az_span _name;
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/az_ulib_descriptor_api.h:44]: Correct, Descriptors are internal.

[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/az_ulib_descriptor_api.h:44]: Just ot give more context, Decriptors are constants, so developers could access it without damage the system, however, the information there can be real or not, depending on the external conditions, so read it directly may result in wrong information.

File: inc/az_ulib_ipc_interface.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+{
+  return vtable->release_interface(interface_handle);
+}
+
+AZ_INLINE az_result azi_ulib_ipc_call(
+    const az_ulib_ipc_vtable* const vtable,
+    az_ulib_ipc_interface_handle interface_handle,
+    az_ulib_capability_index command_index,
+    az_ulib_model_in model_in,
+    az_ulib_model_out model_out)
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/az_ulib_ipc_interface.h:123]: Yes, it shall be a pointer, but az_ulib_model_out is already a pointer:
```
typedef void* az_ulib_model_out;
```

**Comment by mamokarz** [comment]
> #ifndef AZ_ULIB_DESCRIPTOR_H

AZ_ULIB_DESCRIPTOR_API_H

---
Refers to: inc/az_ulib_descriptor_api.h:5 in 1cf0eb7. [](commit_id = 1cf0eb748edc07629b79be77e19903cbbef23d19, deletion_comment = False)

**Comment by mamokarz** [comment]
> #ifndef AZ_ULIB_IPC_H

AZ_ULIB_IPC_API_H

---
Refers to: inc/az_ulib_ipc_api.h:14 in 1cf0eb7. [](commit_id = 1cf0eb748edc07629b79be77e19903cbbef23d19, deletion_comment = False)

File: inc/az_ulib_ipc_interface.h [hunk 5]
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#ifndef AZ_ULIB_IPC_VTABLE_H
```
Comment (mamokarz) [inline inc/az_ulib_ipc_interface.h:5]: >AZ_ULIB_IPC_VTABLE_H [](http://example.com/codeflow?start=8&length=20)

AZ_ULIB_IPC_INTERFACE_H

File: inc/az_ulib_ipc_interface.h [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+#include <stddef.h>
+#include <stdint.h>
+#else
+#include <cstddef>
+#include <cstdint>
+#endif /* __cplusplus */
+
+#include "azure/core/_az_cfg_prefix.h"
+
+typedef struct
```
Comment (mamokarz) [inline inc/az_ulib_ipc_interface.h:23]: >struct [](http://example.com/codeflow?start=8&length=6)

Add documentation.

File: inc/az_ulib_ipc_interface.h [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      az_span model_in_span,
+      az_span* model_out_span);
+
+  az_result (*query)(az_span query, az_span* result, uint32_t* continuation_token);
+
+  az_result (*query_next)(uint32_t* continuation_token, az_span* result);
+
+} az_ulib_ipc_vtable;
+
+AZ_INLINE az_result azi_ulib_ipc_publish(
```
Comment (mamokarz) [inline inc/az_ulib_ipc_interface.h:68]: >AZ_INLINE [](http://example.com/codeflow?start=0&length=9)

Add AZ_NODISCARD to these APIs.

File: inc/az_ulib_ipc_interface.h [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+  az_result (*release_interface)(az_ulib_ipc_interface_handle interface_handle);
+
+  az_result (*call)(
+      az_ulib_ipc_interface_handle interface_handle,
+      az_ulib_capability_index command_index,
+      az_ulib_model_in model_in,
+      az_ulib_model_out model_out);
+
+  az_result (*call_w_str)(
```
Comment (mamokarz) [inline inc/az_ulib_ipc_interface.h:56]: >call_w_str [](http://example.com/codeflow?start=14&length=10)

call_with_str

File: src/az_ulib_ipc/az_ulib_ipc_query_interface.c [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+#define IPC_QUERY_1_INTERFACE_NAME "ipc_" QUERY_1_INTERFACE_NAME
+
+static az_result query_1_query_concrete(az_ulib_model_in model_in, az_ulib_model_out model_out)
+{
+  const query_1_query_model_in* const in = (const query_1_query_model_in* const)model_in;
+  query_1_query_model_out* out = (query_1_query_model_out*)model_out;
+  return az_ulib_ipc_query(in->query, out->result, &(out->continuation_token));
+}
+
+inline uint32_t model_out_span_min_size(void)
```
Comment (mamokarz) [inline src/az_ulib_ipc/az_ulib_ipc_query_interface.c:34]: >inline [](http://example.com/codeflow?start=0&length=6)

AZ_INLINE

File: samples/ipc_call_interface/producers/cipher_v2i1/cipher_v2i1.c [hunk 10]
```diff
@@ -51,9 +52,34 @@ static inline uint32_t next_key_pos(uint32_t cur)
   return next;
 }
 
-static uint32_t encoded_len(uint32_t size)
+static int32_t encoded_len(az_span src)
 {
-  return (size == 0) ? (0) : (((((size - 1) / 3) + 1) * 4) + 1);
+  int32_t size = az_span_size(src);
+  return (size == 0) ? (0) : (((((size - 1) / 3) + 1) * 4) + 1) + 1;
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply samples/ipc_call_interface/producers/cipher_v2i1/cipher_v2i1.c:58]: Sure.

**Comment by mamokarz** [comment]
That is fine, thanks for the help.

---
In reply to: [668205104](https://github.com/Azure/azure-ulib-c/pull/69#issuecomment-668205104) [](http://example.com/codeflow?ancestors=668205104,668205104)

File: inc/az_ulib_descriptor_api.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (danewalton) [reply inc/az_ulib_descriptor_api.h:44]: nit: maybe consider wrapping them in a similar way that the embedded does in an `_internal` struct:
```c


typedef struct az_ulib_capability_descriptor_tag
{
    struct {
        const az_span _name;
    } _internal;
}
```

https://github.com/Azure/azure-sdk-for-c/blob/e82661b4844197f4a3a216b26e11cfae917eaedf/sdk/inc/azure/iot/az_iot_hub_client.h#L63-L71

File: src/az_ulib_ipc/az_ulib_ipc_query_interface.c (hunk 9 above)
Comment (mamokarz) [inline src/az_ulib_ipc/az_ulib_ipc_query_interface.c:34]: >uint32_t [](http://example.com/codeflow?start=7&length=8)

int32_t

File: src/az_ulib_ipc/az_ulib_ipc_query_interface.c [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        AZ_ULIB_THROW_IF_AZ_ERROR(
+            az_span_atou32(jr.token.slice, &(next_model_in.continuation_token)));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the next_model_out.
+    char dest_buffer[200];
```
Comment (mamokarz) [inline src/az_ulib_ipc/az_ulib_ipc_query_interface.c:156]: >dest_buffer [](http://example.com/codeflow?start=9&length=11)

Use model_out_span instead, as in the query.

File: src/az_ulib_ipc/az_ulib_ipc_query_interface.c [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        query_model_in.query
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the query_model_out.
+    az_span dest_span = az_span_slice_to_end(*model_out_span, model_out_span_min_size() + 1);
```
Comment (mamokarz) [inline src/az_ulib_ipc/az_ulib_ipc_query_interface.c:110]: >az_span dest_span = az_span_slice_to_end(*model_out_span, model_out_span_min_size() + 1); [](http://example.com/codeflow?start=4&length=89)

Windows build in the gate is not happy with it.

File: inc/az_ulib_descriptor_api.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/az_ulib_descriptor_api.h:44]: In this case, we should do the same for `az_ulib_interface_descriptor`

File: inc/az_ulib_capability_api.h [hunk 13]
```diff
@@ -123,25 +124,73 @@ typedef az_result (*az_ulib_capability_cancellation_callback)(
 typedef az_result (
     *az_ulib_capability_command)(az_ulib_model_in model_in, az_ulib_model_out model_out);
 
+/**
+ * @brief       Call a capability in the interface using strings in `az_span`.
+ *
+ * This type defines the signature for the synchronous commands that will be published in an IPC
+ * interface. A synchronous command is the one which runs in the same call stack as the caller.
```
Comment (cindydeng1998) [inline inc/az_ulib_capability_api.h:131]: >is the one which [](http://example.com/codeflow?start=36&length=16)

 A synchronous command runs in the same call stack as the caller

File: inc/az_ulib_ipc_api.h [hunk 14]
```diff
@@ -56,11 +57,6 @@ typedef struct az_ulib_ipc_tag
   } _internal;
 } az_ulib_ipc;
 
-/**
- * @brief Interface handle.
- */
-typedef void* az_ulib_ipc_interface_handle;
```
Comment (cindydeng1998) [inline inc/az_ulib_ipc_api.h:62]: why was this moved to the descriptor api?

File: inc/az_ulib_ipc_api.h [hunk 15]
```diff
@@ -292,13 +326,92 @@ AZ_NODISCARD az_result az_ulib_ipc_release_interface(az_ulib_ipc_interface_handl
...
+ *                                  different than 0, means that a call to
+ *                                  az_ulib_ipc_query_next() shall return more information.
+ *
+ * @pre     IPC shall already been initialized.
+ * @pre     \p result shall be a valid az_span with at least 1 position.
+ * @pre     \p continuation_token shall be different than `NULL`.
+ *
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                      If the succeeded and the result and continuation have
+ *                                      valid information.
```
Comment (cindydeng1998) [inline inc/az_ulib_ipc_api.h:388]: nit: If the query call succeeded and the result or continuation has valid information.

File: inc/az_ulib_capability_api.h (hunk 13 above)
Comment (mamokarz) [reply inc/az_ulib_capability_api.h:131]: Yes

File: inc/az_ulib_ipc_api.h (hunk 14 above)
Comment (mamokarz) [reply inc/az_ulib_ipc_api.h:62]: Because packages cannot include az_ulib_ipc_api.h, It shall include az_ulib_ipc_interface.h instead, but packages needs this definiction, so, the common point is az_ulib_descriptor_api.h

File: inc/az_ulib_ipc_api.h (hunk 15 above)
Comment (mamokarz) [reply inc/az_ulib_ipc_api.h:388]: same for query next. thanks!

File: inc/az_ulib_ipc_api.h [hunk 16]
```diff
@@ -287,18 +321,97 @@ AZ_NODISCARD az_result az_ulib_ipc_release_interface(az_ulib_ipc_interface_handl
...
+ * @pre     \p continuation_token shall not be 'NULL'.
+ *
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                      If the succeeded and the result and continuation have
+ *                                      valid information.
+ *  @retval #AZ_ULIB_EOF                If there is no more information to return in this query.
+ *  @retval #AZ_ERROR_ITEM_NOT_FOUND    If the target command does not exist.
+ */
+AZ_NODISCARD az_result
+az_ulib_ipc_query(az_span query, az_span* result, uint32_t* continuation_token);
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:393]: >az_ulib_ipc_query [](http://example.com/codeflow?start=0&length=17)

Missing AZ_ERROR_NOT_SUPPORTED. In the query next as well.

File: inc/az_ulib_ipc_api.h [hunk 17]
```diff
@@ -287,18 +321,97 @@ AZ_NODISCARD az_result az_ulib_ipc_release_interface(az_ulib_ipc_interface_handl
...
+ *
+ * @pre     IPC shall already been initialized.
+ * @pre     \p result shall be a valid az_span with at least 1 position.
+ * @pre     \p continuation_token shall not be 'NULL'.
+ *
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                      If the succeeded and the result and continuation have
+ *                                      valid information.
+ *  @retval #AZ_ULIB_EOF                If there is no more information to return in this query.
+ *  @retval #AZ_ERROR_ITEM_NOT_FOUND    If the target command does not exist.
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:390]: >AZ_ERROR_ITEM_NOT_FOUND [](http://example.com/codeflow?start=13&length=23)

Remove

File: src/az_ulib_ipc/az_ulib_ipc_query_interface.c (hunk 12 above)
Comment (mamokarz) [reply src/az_ulib_ipc/az_ulib_ipc_query_interface.c:110]: _az_cfg.h disable this warning.

Comment (mamokarz) [reply src/az_ulib_ipc/az_ulib_ipc_query_interface.c:110]: disable this warning in teh compile_options.

File: inc/az_ulib_capability_api.h (hunk 13 above)
Comment (smcd253) [reply inc/az_ulib_capability_api.h:131]: nit^^

File: tests/src/az_ulib_test_my_interface.c [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+        model_in.capability = (uint8_t)val;
+      }
+      else if (az_json_token_is_text_equal(
+                   &jr.token, AZ_SPAN_FROM_STR(MY_INTERFACE_MY_COMMAND_MAX_SUM_NAME)))
+      {
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_token_get_uint32(&jr.token, &model_in.max_sum));
+      }
+      else if (az_json_token_is_text_equal(
+                   &jr.token, AZ_SPAN_FROM_STR(MY_INTERFACE_MY_COMMAND_DESCRIPTOR_NAME)))
```
Comment (mamokarz) [inline tests/src/az_ulib_test_my_interface.c:163]: >MY_INTERFACE_MY_COMMAND_DESCRIPTOR_NAME [](http://example.com/codeflow?start=47&length=39)

Trow not support. In a generic platform, we cannot accept local pointer over JSON.

File: tests/src/az_ulib_test_my_interface.c [hunk 19]
```diff
@@ -0,0 +1,428 @@
...
+        model_in.descriptor = (az_ulib_interface_descriptor*)val;
+      }
+      else if (az_json_token_is_text_equal(
+                   &jr.token, AZ_SPAN_FROM_STR(MY_INTERFACE_MY_COMMAND_WAIT_POLICY_MS_NAME)))
+      {
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_token_get_uint32(&jr.token, &model_in.wait_policy_ms));
+      }
+      else if (az_json_token_is_text_equal(
+                   &jr.token, AZ_SPAN_FROM_STR(MY_INTERFACE_MY_COMMAND_HANDLE_NAME)))
```
Comment (mamokarz) [inline tests/src/az_ulib_test_my_interface.c:177]: >MY_INTERFACE_MY_COMMAND_HANDLE_NAME [](http://example.com/codeflow?start=47&length=35)

Trow not support. In a generic platform, we cannot accept local pointer over JSON.

File: inc/az_ulib_ipc_api.h [hunk 20]
```diff
@@ -214,6 +219,35 @@ AZ_NODISCARD az_result az_ulib_ipc_try_get_interface(
...
+ * @brief   Try get a capability index in the interface by the name from the IPC.
+ *
+ * This API tries to find an capability that fits the provided name in the interface. It returns
+ * the capability index.
+ *
+ * @param[in]   interface_handle    The #az_ulib_ipc_interface_handle with the interface handle.
+ *                                  It cannot be `NULL`. Call az_ulib_ipc_try_get_interface() to
+ *                                  get the interface handle.
+ * @param[in]   name                The `az_span` with the interface name.
+ * @param[out]  capability_index    The #az_ulib_capability_index* with the memory to store
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:232]: this was already present before this PR, correct? not entirely sure what it indicates.

File: inc/az_ulib_ipc_api.h [hunk 21]
```diff
@@ -287,18 +321,98 @@ AZ_NODISCARD az_result az_ulib_ipc_release_interface(az_ulib_ipc_interface_handl
...
     az_ulib_model_in model_in,
     az_ulib_model_out model_out);
 
+/**
+ * @brief   Synchronously Call a published procedure using string models.
+ *
+ * @param[in]   interface_handle    The #az_ulib_ipc_interface_handle with the interface handle.
+ *                                  It cannot be `NULL`. Call az_ulib_ipc_try_get_interface() to
+ *                                  get the interface handle.
+ * @param[in]   command_index       The #az_ulib_capability_index with the command index. Call
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:343]: is this like a priority queue?

File: samples/ipc_call_interface/consumers/my_consumer.c [hunk 22]
```diff
@@ -4,13 +4,21 @@
...
 #include "az_ulib_result.h"
+#include "azure/az_core.h"
 #include "wrappers/cipher_1_wrapper.h"
 #include <inttypes.h>
 #include <stdint.h>
 #include <stdio.h>
 
 static az_ulib_ipc_interface_handle _cipher_1;
-#define BUFFER_SIZE 100
+#define BUFFER_SIZE 200
```
Comment (smcd253) [inline samples/ipc_call_interface/consumers/my_consumer.c:14]: can you add a description for how you landed on this buffer size so we know how to tweak it if need be?

File: samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c [hunk 23]
```diff
@@ -23,58 +23,124 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        encrypt_model_in.src
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the encrypt_model_out.
+    char dest_buffer[200];
```
Comment (smcd253) [inline samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c:60]: can you use a #define for the size here?

File: samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c [hunk 24]
```diff
@@ -23,58 +23,124 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        decrypt_model_in.src
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the decrypt_model_out.
+    char dest_buffer[200];
```
Comment (smcd253) [inline samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c:112]: same as above

File: samples/ipc_call_interface/common/cipher_1_model.h [hunk 25]
```diff
@@ -7,11 +7,12 @@
  * modified.
  ********************************************************************/
 
-#ifndef CIPHER_1_INTERFACE_H
-#define CIPHER_1_INTERFACE_H
+#ifndef CIPHER_1_MODEL_H
```
Comment (smcd253) [inline samples/ipc_call_interface/common/cipher_1_model.h:10]: wasn't the plan to rename this to keyvault?

File: samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c (hunk 23 above)
Comment (mamokarz) [reply samples/ipc_call_interface/producers/cipher_v1i1/interfaces/cipher_v1i1_interface.c:60]: I'll actually use model_out_span as buffer instead of create a new one in the stack. 

I'll do the same for decript as well.

File: samples/ipc_call_interface/common/cipher_1_model.h (hunk 25 above)
Comment (mamokarz) [reply samples/ipc_call_interface/common/cipher_1_model.h:10]: Not in this sample, let's keep it a simple cipher.

**Review by cindydeng1998 — APPROVED** [review]
🌞

File: tests/tests_e2e/az_ulib_ipc_e2e/az_ulib_ipc_e2e.c [hunk 26]
```diff
@@ -657,6 +493,178 @@ static void az_ulib_ipc_e2e_call_sync_command_in_multiple_threads_and_unpublish_
...
+  az_ulib_ipc_interface_handle query_handle;
+  assert_int_equal(
+      az_ulib_ipc_try_get_interface(
+          AZ_SPAN_FROM_STR(IPC_QUERY_1_INTERFACE_NAME),
+          QUERY_1_INTERFACE_VERSION,
+          AZ_ULIB_VERSION_EQUALS_TO,
+          &query_handle),
+      AZ_OK);
+
+  uint8_t buf[170];
```
Comment (smcd253) [inline tests/tests_e2e/az_ulib_ipc_e2e/az_ulib_ipc_e2e.c:637]: Can this buf size be a #define? or have a comment describing why it's this size?

Comment (mamokarz) [reply tests/tests_e2e/az_ulib_ipc_e2e/az_ulib_ipc_e2e.c:637]: Better to just add a comment here, this number is important for this paticular test.

File: inc/az_ulib_capability_api.h (hunk 13 above)
Comment (smcd253) [reply inc/az_ulib_capability_api.h:131]: right, it was just a grammar correction. ".... command runs in the same call stack..."

File: inc/az_ulib_ipc_interface.h [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+#include <cstddef>
+#include <cstdint>
+#endif /* __cplusplus */
+
+#include "azure/core/_az_cfg_prefix.h"
+
+/**
+ * @brief Vtable to IPC APIs.
+ *
+ * Set of pointers to the IPC APIs used to expose the IPC APIs to Modules that cannot be statically
```
Comment (smcd253) [inline inc/az_ulib_ipc_interface.h:26]: nit: "... uses to expose..."

File: inc/az_ulib_ipc_api.h [hunk 28]
```diff
@@ -287,18 +321,98 @@ AZ_NODISCARD az_result az_ulib_ipc_release_interface(az_ulib_ipc_interface_handl
...
+/**
+ * @brief   Query next IPC information.
+ *
+ * Retrieves the next information from the IPC using the continuation token.
+ *
+ * @param[in]   continuation_token  The pointer to `uint32_t` with the current continuation
+ *                                  token and where it will return the next continuation token.
+ * @param[in]   result              The `az_span` with the buffer to return the query result.
+ *
+ * @pre     IPC shall already been initialized.
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:404]: nit: "IPC shall already be initialized"

File: samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c [hunk 29]
```diff
@@ -23,58 +23,124 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        decrypt_model_in.src
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the decrypt_model_out.
+    char dest_buffer[200];
```
Comment (mamokarz) [inline samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c:112]: >dest_buffer [](http://example.com/codeflow?start=9&length=11)

Remove this buffer

File: samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c [hunk 30]
```diff
@@ -23,58 +23,124 @@
...
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        encrypt_model_in.src
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the encrypt_model_out.
+    char dest_buffer[200];
```
Comment (mamokarz) [inline samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c:60]: >dest_buffer [](http://example.com/codeflow?start=9&length=11)

Remove this buffer.

File: samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c (hunk 29 above)
Comment (smcd253) [inline samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c:112]: nit: buffer size description

Comment (mamokarz) [reply samples/ipc_call_interface/producers/cipher_v2i1/interfaces/cipher_v2i1_interface.c:112]: This buffer was removed.

File: samples/ipc_call_interface/consumers/my_consumer.c [hunk 31]
```diff
@@ -36,6 +44,22 @@ void my_consumer_create(void)
 {
   (void)printf("Create my consumer...\r\n");
 
+  az_result result;
+  uint32_t continuation_token = 0;
+  uint8_t buf[100];
```
Comment (smcd253) [inline samples/ipc_call_interface/consumers/my_consumer.c:49]: nit: explain buffer size

Comment (mamokarz) [reply samples/ipc_call_interface/consumers/my_consumer.c:49]: Use BUFFER_SIZE here as well.

File: inc/az_ulib_capability_api.h [hunk 32]
```diff
@@ -123,25 +124,73 @@ typedef az_result (*az_ulib_capability_cancellation_callback)(
...
+ * released as soon as the command returns.
+ *
+ * As a standard, the synchronous command shall return #az_result. If the command needs to
+ * return anything else, the data shall be stored in the `model_out`.
+ *
+ * Both `model_in` and `model_out` shall be defined as part of the interface definition, and
+ * both shall be strings with JSON inside.
+ *
+ * @param[in]   model_in_span   The `az_span` that contains a JSON with the input arguments for
+ *                              the command. It may be an empty JSON `{}`, the IPC will not do any
```
Comment (smcd253) [inline inc/az_ulib_capability_api.h:142]: ", the IPC will not validate it."

File: inc/az_ulib_capability_api.h [hunk 33]
```diff
@@ -123,25 +124,73 @@ typedef az_result (*az_ulib_capability_cancellation_callback)(
...
+ *                              validation on it. The command itself shall implement the JSON
+ *                              parser with any needed validation.
+ * @param[out]  model_out_span  The `az_ulib_model_out` that contains the memory to store the
+ *                              JSON with the output arguments from the command. It may be `NULL`,
+ *                              the IPC will not do any validation on it. The command itself shall
```
Comment (smcd253) [inline inc/az_ulib_capability_api.h:147]: nit: same as above

**Review by smcd253 — APPROVED** [review]
lgtm
